File: 6. Zigzag Conversion/zigzag_conversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows == 1 or numRows >= len(s):
            return s

        rows = [''] * numRows
        curRow, step = 0, -1  

        for c in s:
            rows[curRow] += c
            if curRow == 0 or curRow == numRows - 1:
                step = -step
            curRow += step

        return ''.join(rows)

File: 6. Zigzag Conversion/ZigzagConversion.cs
public class Solution {
    public string Convert(string s, int numRows) {
        if (numRows == 1 || s.Length <= numRows) {
            return s;
        }

        string[] rows = new string[Math.Min(numRows, s.Length)];
        int curRow = 0;
        bool goingDown = false;

        foreach (char c in s) {
            rows[curRow] += c;
            if (curRow == 0 || curRow == numRows - 1) {
                goingDown = !goingDown;
            }
            curRow += goingDown ? 1 : -1;
        }

        string result = string.Join("", rows);
        return result;
    }
}